### DISCUSSION TITLE
Use React Hook Form With MUI DatePicker Component In Typescript

### DISCUSSION THREAD

**Discussion by akashshyamdev** [react-hook-form/react-hook-form#6986] [Q&A]
Hey Guys,

I'm using React Hook Form & the MUI `DatePicker` Component with Typescript. I have a `DateField` component which is a reusable component containing the `DatePicker`.

```jsx
// custom wrapper component for DatePicker
<DateField<ReceptionInvoice>
  label="Issued Date"
  control={control}
  name={"invoiceDate"}
  options={{
    required: "This field is required",
  }}
  error={errors.invoiceDate}
  format="MM/dd/yyyy"
  renderInput={(params) => (
    <TextField {...register("invoiceDate")} {...params} />
  )}
/>
```

The `TextField` component is a wrapper for the MUI `textField` component. You can find the code below.

The `DateField` component has the following code:

```jsx
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import DatePicker, { DatePickerProps } from "@mui/lab/DatePicker";
import LocalizationProvider from "@mui/lab/LocalizationProvider";
import { makeStyles } from "@mui/styles";
import React from "react";
import { Control, Controller, FieldValues } from "react-hook-form";
import { ENOFieldOverrides, ENOFieldProps } from "./EnoFieldUtils";
import ENOTooltip from "./ENOTooltip";

export interface ENODateFieldProps<T extends FieldValues>
	extends Omit<ENOFieldProps<T>, "register">,
		Omit<DatePickerProps, ENOFieldOverrides | "value" | "onChange"> {
	control: Control<T, object>;
	format: string;
}

const useStyles = makeStyles({
	fullWidth: {
		width: "100% !important",
	},
});

export default function ENODateField<T extends FieldValues>({
	name,
	control,
	options,
	error,
	tooltipText,
	format,
	...rest
}: ENODateFieldProps<T>) {
	return (
		<ENOTooltip text={tooltipText}>
			<LocalizationProvider dateAdapter={AdapterDateFns}>
				<Controller
					name={name}
					control={control}
					render={({ field: { ref, ...fieldRest } }) => (
						<DatePicker
							// className={classes.fullWidth}
							{...fieldRest}
							{...rest}
							inputFormat={format}
						/>
					)}
				/>
			</LocalizationProvider>
		</ENOTooltip>
	);
}
```

Custom `TextField` component:
```
import { makeStyles, OutlinedTextFieldProps, TextField, TextFieldProps, Theme } from "@mui/material";
import React, { useMemo } from "react";
import { FieldValues } from "react-hook-form";
import { ENOFieldOverrides, ENOFieldProps } from "./EnoFieldUtils";
import ENOTooltip from "./ENOTooltip";

export interface ENOTextFieldProps<T extends FieldValues>
    extends ENOFieldProps<T>,
        Omit<OutlinedTextFieldProps, ENOFieldOverrides> {
    variant?: TextFieldProps["variant"];
}

export default function ENOTextField<T extends FieldValues>({
    name,
    register,
    options,
    variant = "outlined",
    error,
    helperText,
    tooltipText,
    ...rest
}: ENOTextFieldProps<T>) {
    return (
        <ENOTooltip text={tooltipText}>
            <TextField
                sx={{
                    width: "100%",
                }}
                {...register(name, options)}
                {...rest}
                helperText={error?.message ? error.message : helperText}
                // variant={variant}
                variant={"outlined"}
                error={Boolean(error)}
            />
        </ENOTooltip>
    );
}
```
UPDATE:

After fiddling with the code a bit, I don't get any error, but the custom `TextField` component is shown without any icon for the DatePicker. However, if I use the MUI `TextField`, then it works perfectly!

Please help me out, looking forward to your reply!

**Comment by bluebill1049** [comment]
worth to look into `<DatePicker className={classes.fullWidth} {...fieldRest} {...rest} />` what are the prop that DatePicker reads, especially onChange, this error looks like the data supplied with onChange is potentially `undefined`

**Reply by akashshyamdev** [reply]
Hey @bluebill1049 

Thanks for replying, after installing date-fns using `yarn add date-fns`, the error disappears, but the output doesn't have a date picker, it only has an input

Output:
![Screenshot 2021-11-07 at 09 39 26](https://user-images.githubusercontent.com/56759828/140631998-52267759-bc63-44c8-9d73-e8afe981462f.png)

Values of `onChange` and `rest`:
![Screenshot 2021-11-07 at 09 38 01](https://user-images.githubusercontent.com/56759828/140631976-d82ca30d-18db-445c-8d10-bac7c027a75c.png)

**Reply by Admiralfeb** [reply]
What are you passing as the `renderInput` for the DatePicker?
are you passing a straight TextField or a function `(params) => <TextField {...params}/>`?

**Reply by akashshyamdev** [reply]
@Admiralfeb @bluebill1049 I've updated the question with some more code & updates :)

**Comment by jrodolforojas** [comment]
Hi @Admiralfeb, it might work for you. Try adding the field property inside DatePicker/TimePicker instead of the TextField of the renderInput.
`
<Controller
        name={property}
        control={control}
        render={({ field }) =>
          <LocalizationProvider dateAdapter={AdapterDateFns}>
            <TimePicker {...field} renderInput={(props) => <TextField margin='normal' {...props} />}/>
          </LocalizationProvider>
        }
      />
`
